### PR TITLE
fix(adapters): bug-fix sweep from upstream 4.27.0 (slack/discord/telegram)

### PR DESCRIPTION
**Pull request by patrick-chinchill**
## Summary

Bundles 4 small upstream bug fixes from `vercel/chat@4.27.0` into one PR. Each is independent, covered by a regression test, and clearly scoped.

> **Note:** This sweep originally also included `vercel/chat#394` (Slack `@mention` regex eats email addresses). That fix was extracted into #91 and merged to main on 2026-05-15. This branch was rebased and the redundant changes dropped — `git diff origin/main` no longer touches `format_converter.py` or `tests/test_slack_format.py`.

| Upstream PR | Adapter | Bug |
|---|---|---|
| [vercel/chat#292](https://github.com/vercel/chat/pull/292) | Slack | Empty `thread_ts` caused `invalid_thread_ts`; `stream()` now degrades to a single `post_message` for empty `thread_ts` instead of raising, so top-level DM streaming replies are no longer silently dropped (#94) |
| [vercel/chat#256](https://github.com/vercel/chat/pull/256) | Discord | Card posts duplicated text (Discord renders both `content` and the card embed) |
| [vercel/chat#395](https://github.com/vercel/chat/pull/395) | Slack | Link previews missing unfurl metadata — `message_changed` events were being filtered |
| [vercel/chat#407](https://github.com/vercel/chat/pull/407) | Telegram | MarkdownV2 rendering — emitted standard Markdown (`**bold**`) instead of legacy MarkdownV2 (`*bold*`), causing `can't parse entities` errors |

### Slack empty `thread_ts` (vercel/chat#292)
Guard `thread_ts` before passing to `client.chat_postMessage` etc. Empty string and missing key both treated as "no thread". For `stream()` specifically, empty `thread_ts` accumulates the iterator and calls `post_message` once instead of raising — top-level Slack DMs encode their thread id with an empty `thread_ts` by design (so `open_dm` subscriptions match incoming DMs), and raising `ValidationError` silently dropped the reply because `Thread._handle_stream` doesn't catch and `_fallback_stream` only triggers when the adapter has no `stream` method at all (#94). Channels with non-empty `thread_ts` still use native streaming. This lets chinchill-api retire its production `MultiTenantSlackAdapter.stream()` subclass override. Regression tests in `tests/test_slack_api.py::TestEmptyThreadTsGuards`.

### Discord card text dedup (vercel/chat#256)
- **Create path**: omit `content` when posting a card. Discord renders both the message content AND the card embed when `content` is set.
- **Edit path**: explicitly send `content: ""` (PATCH preserves omitted fields). Without this, leftover text from a previous edit persists alongside the card.
- Tests in `test_discord_extended.py` cover both paths.

### Slack link unfurl enrichment (vercel/chat#395)
- `message_changed` is no longer in `_IGNORED_SUBTYPES` — routed to new `_handle_message_changed`.
- Slack delivers unfurled attachments via a separate `message_changed` event 100-2000ms after the original. Added `_UNFURL_WAIT_MS = 2000` poll window with `_UNFURL_POLL_MS = 150` cadence so the message handler sees enriched links.
- Per-event link cache (1h TTL) so we don't double-resolve.
- Tests in `test_slack_webhook.py`.

### Telegram MarkdownV2 rewrite (vercel/chat#407)
- Replaces legacy `Markdown` parse_mode with `MarkdownV2`. Standard markdown emitted `**bold**`; Telegram MarkdownV2 wants `*bold*`.
- Three escape contexts: normal text (18 chars: `_*[]()~\`>#+-=|{}.!\`), code blocks (only `` ` `` and `\`), inline-link URLs (only `)` and `\`).
- Tests in `test_telegram_format.py` cover headings, bold, italic, strikethrough, code blocks, inline code, links, lists, blockquotes, escape edge cases.

## Test plan

- [x] `uv run ruff check src/ tests/ scripts/` — clean
- [x] `uv run ruff format --check src/ tests/ scripts/` — clean
- [x] `uv run python scripts/audit_test_quality.py` — 0 hard failures (39 pre-existing warnings unchanged)
- [x] `uv run pytest tests/ --tb=short -q` — **3702 pass, 2 skipped, 1 pre-existing failure** (`tests/test_github_webhook.py::TestGitHubAdapterConstructor::test_throws_when_no_auth`, unrelated to this PR — fails on main today). _Pre-rebase counts._
- [x] Post-rebase narrow check: `pytest tests/test_slack_format.py -q` → 91 passed.
- [x] Post-fold-in (#94) narrow check: `pytest tests/test_slack_api.py tests/test_slack_adapter.py tests/test_slack_format.py tests/test_slack_extended.py -q` → 228 passed.

## Upstream refs

- vercel/chat#292 (`53c6b68`)
- vercel/chat#256 (`7e5b447`)
- vercel/chat#395 (`ded6f78`)
- vercel/chat#407 (`b9a1961`)

Closes #94.

https://claude.ai/code/session_01FyMxQn2BEAzmwKS1GZczKj

---
_Generated by [Claude Code](https://claude.ai/code/session_01FyMxQn2BEAzmwKS1GZczKj)_

## Summary by CodeRabbit

* **Bug Fixes**
  * Discord: Card messages no longer show duplicate plain-text; edits reliably clear previous text.
  * Telegram: Truncation now preserves valid MarkdownV2 entities to avoid rejected or broken messages.

* **New Features**
  * Slack: Link previews are enriched from message edits (better unfurl/title/description/image handling) and streaming for direct/top-level messages falls back to reliable post-based accumulation.
  * Telegram: Cards and rich content now render using MarkdownV2 for improved formatting.

[Review Change Stack](https://app.coderabbit.ai/change-stack/Chinchill-AI/chat-sdk-python/pull/89?utm_source=github_walkthrough&utm_medium=github&utm_campaign=change_stack)

### CODE REVIEW

**Review by patrick-chinchill — COMMENTED** [review]
## Review of bug-fix sweep from upstream `chat@4.27.0`

Read PR + 2 follow-up commits (`3bd5149` CodeQL test fix, `0de162c` unfurl logging) against upstream `f55378a` (`chat@4.27.0`) commits `c26ee6c`, `53c6b68`, `7e5b447`, `ded6f78`, `b9a1961`. Applied porting hazards (`docs/UPSTREAM_SYNC.md`) + adversarial checks (`docs/SELF_REVIEW.md`). Ran the touched test files locally — all 443 pass.

---

### 1. Slack `@mention` regex (vercel/chat#394) — ✅ Looks good

Python regex `(?<![<\w])@(\w+)` is character-for-character identical to TS `BARE_MENTION_REGEX`. Hoisted to module-level constant matching upstream. Both call sites updated. Six regression tests cover email-not-mangled, mailto-preserved, leading `@user` still wrapped, `@@user` adversarial, dotted-subdomain emails. Test docstrings include "What to fix if this fails:" line per CLAUDE.md principle #8.

🔵 **Nit**: `@user@example.com` (mention immediately followed by an email-shaped tail) is not in the test sweep. Verified manually: produces `<@user>@example.com` (the `@user` is wrapped, the email tail is preserved). That's parity with upstream TS but worth pinning a test on.

---

### 2. Slack empty `thread_ts` guard (vercel/chat#292) — ✅ Looks good

`thread_ts or None` normalization at every Slack API call site (`post_message`, `post_ephemeral`, `schedule_message`, `upload_files`, `start_typing`). `stream` correctly raises `ValidationError` early. Adversarial `thread_ts == "0"` test is present and passes (`"0"` is truthy in Python so it survives the `or None` collapse — the test pins this behavior). Truthiness hazard #1 from `UPSTREAM_SYNC.md` correctly applied.

---

### 3. Discord card text dedup (vercel/chat#256) — ✅ Looks good

Both Python sites updated:
- `post_message`: omits `content` when card present (parity with TS).
- `edit_message`: explicitly sets `content = ""` (parity with TS — comment correctly notes "Discord PATCH preserves omitted fields").

The third TS site (`postChannelMessage`) is not ported — confirmed the Python Discord adapter has no `post_channel_message` method (declared on `Adapter` protocol but unimplemented for Discord; only Google Chat / Telegram implement it). This is correct, not a port gap. Tests cover create, edit, and the "card with no text content" adversarial. The `assert "content" in payload` check on edit is the load-bearing assertion — correct.

---

### 4. Slack link unfurl enrichment (vercel/chat#395) — 🟡 Medium + ✅

Faithful port of `_handle_message_changed`, `_extract_links` (inline attachments), `_enrich_links` (poll loop with 2000ms / 150ms cadence), 1h cache TTL. Trailing-slash normalization in both directions. Async write in `_handle_message_changed` is correctly tracked via `create_task` + `add_done_callback` (hazard #5 applied). Follow-up commit `0de162c` added the requested state-backend error logging in `_enrich_links` — verified at `src/chat_sdk/adapters/slack/adapter.py:1843-1847`.

🟡 **Medium — merge precedence diverges from TS**: `_merge_unfurl_into_preview` (`adapter.py:1748-1755`) preserves `preview.<field>` if non-`None`, falling back to unfurl. TS does `{ ...link, ...unfurl }` — unfurl OVERRIDES the link's existing fields (except `title`, which is short-circuited above). For `description` / `image_url` / `site_name`, behavior diverges when the preview already has a value: Python keeps the preview value, TS overwrites with the (possibly fresher) unfurl. In practice rare since `_create_link_preview` returns blank fields, but it's a silent semantic divergence not flagged in `UPSTREAM_SYNC.md` non-parity table. Either match TS spread-overwrite semantics, or document the divergence + add a regression test.

🔵 **Nit**: TS `url.replace(TRAILING_SLASH_PATTERN, "")` strips a single trailing `/`; Python `_TRAILING_SLASH_PATTERN.sub("", url)` strips ALL (re.sub default count=0). For `https://x.com//` they differ. Use `.sub("", url, count=1)` for parity.

🔵 **Nit**: 1h cache TTL — if a message is edited multiple times, each `message_changed` overwrites prior unfurls (parity with TS `state.set`). No merge across edits, no reset. Worth one regression test pinning the overwrite semantics.

🔵 **Nit**: `_enrich_links` makes the message-parse path block up to 2000ms inside the chat factory. This serializes one webhook for 2s when no `message_changed` is in flight (e.g. message with no real unfurls). Parity with TS, but worth noting in a comment that the 2000ms is a per-message worst case.

---

### 5. Telegram MarkdownV2 (vercel/chat#407) — 🔴 Critical

Renderer port is faithful. 18 special chars in normal text, 2 in code blocks, 2 in link URLs — all match TS regex character-for-character. `from_markdown(card_to_fallback_text(card))` correctly handles the Python-specific `**title**` shape (Python `card_to_fallback_text` is hard-coded to `**`, no `boldFormat` arg needed). Headings → bold, lists with escaped `\-` and `\1.`, blockquote `>` per line, table preprocessed → ASCII code block. `resolve_parse_mode` correctly returns MarkdownV2 for `{ast}`/`{markdown}`/cards/JSX, `None` for plain str + `{raw}`.

🔴 **Critical — truncation produces invalid MarkdownV2**: TS commit b9a1961 explicitly added `truncateForTelegram` + `trimToMarkdownV2SafeBoundary` because the naive `slice + "..."` fails on MarkdownV2:
- `.` is reserved → bare `"..."` is a parse error
- The slice can leave an orphan trailing `\` (escapes the appended ellipsis or nothing)
- The slice can cut through a paired entity (`*bold*`, `` `code` ``) leaving it unclosed

Python `truncate_message` (`src/chat_sdk/adapters/telegram/adapter.py:1739-1741`) still calls `_truncate_to_utf16(text, TELEGRAM_MESSAGE_LIMIT)` with default `ellipsis="..."` — none of the three failure modes are guarded. `truncate_caption` has the same gap. Any MarkdownV2 message that exceeds 4096 chars (or 1024 for captions) will trigger `Bad Request: can't parse entities` — exactly the bug class this PR claims to fix. Existing length tests only check `len(result) <= 4096`, not parser validity.

Port `truncateForTelegram` + `trimToMarkdownV2SafeBoundary` + `findUnescapedPositions` + `endsWithOrphanBackslash`. Add the 8 length-limit tests from upstream (`*` crosses 4096, `` ` `` crosses 4096, orphan `\` at boundary).

🔵 **Nit**: TS renderer throws on unknown node types via `node satisfies never`. Python falls back to "render children else escape value else empty" — silently swallows new mdast node kinds instead of failing the build. Acceptable trade-off (Python has no exhaustiveness check), but worth a comment.

🔵 **Nit**: Test file uses generic class `TestMarkdownV2Escaping` for all 11 tests but doesn't cover URL with `(` (escape rule says only `)` and `\` are escaped — a URL with unbalanced `(` will likely break Telegram's parser).

---

### Cross-cutting

- ✅ All 5 fixes have regression tests with `What to fix if this fails:` docstrings (CLAUDE.md principle #8).
- ✅ AsyncMock used correctly throughout (no MagicMock-where-AsyncMock bugs surfaced — principle #2).
- ✅ Truthiness hazard correctly applied (`thread_ts or None`, with `"0"` adversarial test).
- ✅ Spawned task in `_handle_message_changed` is tracked + error-handled (hazard #5).
- 🟡 The Telegram truncation gap is a regression-class issue carried over from the pre-MarkdownV2 code — the parse_mode change *widens* it (legacy Markdown was permissive; MarkdownV2 is strict). Recommend landing the truncation port in this PR or as an immediate follow-up before any 0.4.27 release.

---

_Posted by an automated reviewer agent. https://claude.ai/code/session_01FyMxQn2BEAzmwKS1GZczKj_

---
_Generated by [Claude Code](https://claude.ai/code)_

**Review by patrick-chinchill — COMMENTED** [review]
## Re-review (post-fix sweep)

Verified the two new commits (`17379c7`, `f21f047`) against upstream `f55378a` (chat@4.27.0) plus the post-4.27.0 streaming-safety patch (`f46a6fb` / chat#446). All 268 tests in the touched files pass; ruff + audit clean.

### Critical fix verification — Telegram MarkdownV2 truncation

- `find_unescaped_positions`, `ends_with_orphan_backslash`, `_trim_to_markdown_v2_safe_boundary`, `truncate_for_telegram` match upstream line-by-line (loop bound `len+1`, marker set, ellipsis constants, slice-then-trim ordering).
- The Python port adds `_find_unescaped_positions_outside_code` — this is a **forward-port of f46a6fb (chat#446)** which is *not* yet in the 4.27.0 tag. Code, comments, and the commit message correctly attribute it. Acceptable port-ahead.
- `truncate_message` / `truncate_caption` plumb `parse_mode` through to `truncate_for_telegram` for `parse_mode == "MarkdownV2"`; plain falls back to UTF-16 with literal `...`. All 3 call sites in `adapter.py` (lines 1009, 1095, 1635 — `post_message`, `edit_message`, `send_document` caption) pass `parse_mode`.
- Adversarial sweep (run live):
  - empty input -> `""` (no crash)
  - input shorter than ellipsis -> returned verbatim
  - code fence with `*args, **kwargs` under limit -> preserved (the `OutsideCode` helper correctly ignores asterisks inside ```` ``` ````)
  - unclosed code fence past cut -> trim back to before the opener; backticks even
  - `[label](http://x.co...` cut mid-URL -> brackets balanced; `(`/`)` not tracked (matches upstream — known shared limitation)
  - all-special escaped (`escape_markdown_v2(".")*200`) -> length OK, no orphan trailing `\`
  - `\\\\` (escaped backslash run) -> trailing parity preserved, no orphan `\`
  - inline `` `code with * inside` `` cut after opener -> drops everything from the unpaired backtick onward (clean)

### Other addressed items (verified)

- 🟡 `_merge_unfurl_into_preview`: now `unfurl wins` for description/image_url/site_name; title short-circuited at `_enrich_links` call site. Regression test `test_enrich_links_unfurl_overrides_existing_description` pins `old -> new`. Matches TS `{...preview, ...unfurl}` semantically (the only divergence — Python keeps preview when unfurl key is `None` rather than overwriting with `undefined` — is benign because `_create_link_preview` never sets these fields).
- 🔵 `@user@example.com` regex test added (`test_slack_format.py:571`).
- 🔵 `_TRAILING_SLASH_PATTERN.sub("", url, count=1)` pinned at both call sites.
- 🔵 Multi-edit cache overwrite test added (`test_message_changed_overwrites_cached_unfurl_not_merge`).
- 🔵 2000ms `_UNFURL_WAIT_MS` latency comment present.
- 🔵 `node satisfies never` divergence comment present in `format_converter.py:159`.
- 🔵 URL-with-`(` regression test added (`test_extract_links_url_with_open_paren_survives_parser`).

### Upstream parity hunt (new findings)

- 🔵 **Nit — Discord `post_channel_message` missing entirely.** Upstream has 3 `cardToFallbackText` -> "don't include text" sites: `postMessage` (l.814), `editMessage` (l.1178), `postChannelMessage` (l.2398). Python only has 2 (`post_message`, `edit_message`); there is no `post_channel_message` method on the Discord adapter. Pre-existing gap, not introduced by this PR. Worth a follow-up issue.
- 🔵 **Nit — `_enrich_links` short-circuit uses `is not None` vs TS truthy.** Python `(link.title is not None) or (link.fetch_message is not None)` differs from TS `l.title || l.fetchMessage`: an empty-string title would short-circuit in Python but not TS. Pre-existing, vanishingly unlikely in practice (Slack doesn't emit `""` titles).
- ✅ Telegram MarkdownV2 renderer covers every node type upstream's `markdown.ts` switch handles (root/paragraph/text/strong/emphasis/delete/inlineCode/code/link/blockquote/list/listItem/heading/thematicBreak/break/image/html/linkReference/imageReference/definition/footnoteDefinition/yaml/footnoteReference/table/tableRow/tableCell). No drop-through gaps.
- ✅ `handle_message_changed`: matches upstream — only attachments-with-`from_url`/`original_url` flow; no other message_changed branches in upstream we missed.

### Re-review verdict: **PASS** — Critical Telegram MarkdownV2 truncation fix is sufficient and faithful to upstream. The two follow-up nits (Discord `post_channel_message`, `_enrich_links` truthy) are pre-existing parity gaps unrelated to this sweep and should not block merge.

_Posted by an automated re-reviewer agent. https://claude.ai/code/session_01FyMxQn2BEAzmwKS1GZczKj_


---
_Generated by [Claude Code](https://claude.ai/code)_

File: src/chat_sdk/adapters/slack/adapter.py [hunk 1]
```diff
@@ -1692,7 +1711,159 @@ def _extract_links(self, event: dict[str, Any]) -> list[LinkPreview]:
...
+            return links
+
+        deadline = time.monotonic() + (_UNFURL_WAIT_MS / 1000.0)
+        state = self._chat.get_state()
+        stored: dict[str, dict[str, str | None]] | None = None
+        while True:
+            try:
+                stored = await state.get(f"slack:unfurls:{message_ts}")
+            except Exception:
+                return links
```
[the remark this answers is not in the document]
Comment (patrick-chinchill) [reply src/chat_sdk/adapters/slack/adapter.py]: Addressed in commit 2d07c6d ("fix(slack): log state-backend failures during unfurl enrichment"). The fallback now logs the exception via `self._logger.warn("Failed to read unfurl data from state", {"error": str(exc), "message_ts": message_ts})` before returning `links` (see `src/chat_sdk/adapters/slack/adapter.py:1902-1906` on the current HEAD `636d1e8`). Closing this thread out.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: src/chat_sdk/adapters/telegram/adapter.py [hunk 2]
```diff
@@ -164,6 +164,183 @@ def _truncate_to_utf16(text: str, limit: int, ellipsis: str = "...") -> str:
...
+
+    Limit is interpreted in Python ``len()`` units (== UTF-16 code units
+    for the BMP, == 1 for astral characters). Telegram's 4096 / 1024
+    caps are documented in UTF-16 code units; for parity with upstream
+    we follow upstream's `string.length` semantics here. Pre-existing
+    UTF-16-aware truncation is reserved for non-MarkdownV2 paths.
+    """
+    is_markdown_v2 = parse_mode == "MarkdownV2"
+
+    if len(text) <= limit:
```
[the remark this answers is not in the document]
Comment (patrick-chinchill) [reply src/chat_sdk/adapters/telegram/adapter.py]: Addressed in commit 112cab2 ("fix(telegram): UTF-16 length + preserve underscores in MarkdownV2 link URLs").

`truncate_for_telegram` now measures both the under-limit gate and the over-limit slice in UTF-16 code units. The new `_slice_to_utf16_units` helper produces a prefix whose UTF-16 length is `<= units` and never splits a non-BMP codepoint mid-surrogate-pair. The MarkdownV2 path therefore matches the plain path's `_utf16_len` accounting, and a 4096-emoji message (8192 UTF-16 units) now truncates instead of slipping past the cap.

Regression test: `tests/test_telegram_api.py::TestTruncateForTelegramUtf16` (4 cases) pins the new accounting, including that surrogate pairs are not split mid-character.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: src/chat_sdk/adapters/telegram/adapter.py [hunk 3]
```diff
@@ -164,6 +164,183 @@ def _truncate_to_utf16(text: str, limit: int, ellipsis: str = "...") -> str:
...
+        return _trim_to_markdown_v2_safe_boundary(text) if is_markdown_v2 else text
```
[the remark this answers is not in the document]
Comment (patrick-chinchill) [reply src/chat_sdk/adapters/telegram/adapter.py:477]: Addressed in commit 112cab2 ("fix(telegram): UTF-16 length + preserve underscores in MarkdownV2 link URLs").

Took Option B (link-destination aware delimiter counting) since Option A would have regressed `test_strips_unpaired_marker_under_limit_streaming_leak` -- under-limit MarkdownV2 still has to go through `_trim_to_markdown_v2_safe_boundary` for streamed chunks with transiently unpaired openers (chat#446 / upstream f46a6fb).

`_find_unescaped_positions_outside_code` now accepts `skip_link_dest=True`: when scanning for the delimiter markers `*`, `_`, `~`, the scanner enters link-destination mode on `](` and exits on the matching `)`, so any of those characters inside the URL are treated as literal text per Telegram's MarkdownV2 spec (only `)` and `\` are special inside `](...)`). The bracket-balancing pass still uses the default (no skip) so unmatched `[` continues to be detected.

Regression test: `tests/test_telegram_api.py::TestTruncateForTelegramLinkDestination` (5 cases): single `_` in URL, multiple `_`, `*` + `~` in query string, link + balanced italic together, and a confirmation that a legitimate unpaired `_` outside the URL is still trimmed.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: src/chat_sdk/adapters/telegram/adapter.py [hunk 4]
```diff
@@ -164,6 +164,183 @@ def _truncate_to_utf16(text: str, limit: int, ellipsis: str = "...") -> str:
...
+                last_unpaired = positions[-1] if positions else len(current)
+                if last_unpaired < min_unsafe_position:
+                    min_unsafe_position = last_unpaired
+
+        open_brackets = _find_unescaped_positions_outside_code(current, "[")
+        close_brackets = _find_unescaped_positions_outside_code(current, "]")
+        if len(open_brackets) > len(close_brackets):
+            last_open = open_brackets[-1] if open_brackets else len(current)
+            if last_open < min_unsafe_position:
+                min_unsafe_position = last_open
```
[the remark this answers is not in the document]
Comment (patrick-chinchill) [reply src/chat_sdk/adapters/telegram/adapter.py:441]: Addressed in commits `112cab2` ("fix(telegram): UTF-16 length + preserve underscores in MarkdownV2 link URLs") and `385983f` ("test(telegram): add coverage for unclosed MarkdownV2 link destinations").

The safe-boundary trimmer now detects inline links with balanced `[]` but a dangling `(` destination via a new helper `_find_unclosed_link_dest_open_bracket` (see `src/chat_sdk/adapters/telegram/adapter.py` on HEAD `f382ea4`). The helper walks the text honoring code-fence/inline-code regions and escape backslashes, tracks `[`...`]` pairing, and when a `]` is immediately followed by `(` it scans forward for an unescaped `)`. If none is found, the corresponding `[` position is returned and the trimmer rolls `min_unsafe_position` back to it.

Six regression tests in `TestTrimUnclosedLinkDestination` pin the behavior:

- bare unclosed destination `[label](https://example.com/very-long-path` -> trimmed to `""`
- prefix-then-unclosed `hello world [label](...` -> preserved as `"hello world "`
- closed link `[ok](https://example.com/foo)` -> unchanged
- `[` inside an inline code span -> ignored (literal text)
- end-to-end via `truncate_for_telegram` for an over-limit message that slices mid-URL -> rolled back before the `[`
- escaped `\)` does not count as a real closer -> trim back to `[`

The Codex underscore-in-URL fix (option B, link-region-aware) is in `112cab2`; this finding's separate `(`-without-`)` detector is layered on top so both under-limit and over-limit cases are handled.

---
_Generated by [Claude Code](https://claude.ai/code/session_01FyMxQn2BEAzmwKS1GZczKj)_

---
_Generated by [Claude Code](https://claude.ai/code)_

File: src/chat_sdk/adapters/telegram/adapter.py [hunk 5]
```diff
@@ -164,6 +164,183 @@ def _truncate_to_utf16(text: str, limit: int, ellipsis: str = "...") -> str:
...
+    if len(text) <= limit:
+        return _trim_to_markdown_v2_safe_boundary(text) if is_markdown_v2 else text
+
+    ellipsis = _MARKDOWN_V2_ELLIPSIS if is_markdown_v2 else _PLAIN_ELLIPSIS
+    sliced = text[: limit - len(ellipsis)]
+
+    if is_markdown_v2:
+        sliced = _trim_to_markdown_v2_safe_boundary(sliced)
+
+    return f"{sliced}{ellipsis}"
```
[the remark this answers is not in the document]
Comment (patrick-chinchill) [reply src/chat_sdk/adapters/telegram/adapter.py:485]: Addressed in commit `112cab2` ("fix(telegram): UTF-16 length + preserve underscores in MarkdownV2 link URLs").

This finding is a duplicate of the Codex P2 #1 (UTF-16 length) finding on this PR -- the fix covers both reports.

`truncate_for_telegram` now measures the MarkdownV2 path in UTF-16 code units to match Telegram's documented 4096/1024 caps. Specifically:

- Under-limit gate switched from `len(text) <= limit` to `_utf16_len(text) <= limit` (see `src/chat_sdk/adapters/telegram/adapter.py:404` on HEAD `f382ea4`).
- Over-limit slice replaced `text[: limit - len(ellipsis)]` with a new `_slice_to_utf16_units(text, limit - _utf16_len(ellipsis))` helper that returns the longest prefix whose UTF-16 length fits the budget without splitting a non-BMP surrogate pair.
- Docstring updated to reflect that the limit is now interpreted in UTF-16 code units (matches upstream JS `string.length` semantics).

Regression tests in `TestTruncateForTelegramUtf16`:

- 4096 grinning-face emoji (each = 2 UTF-16 units, total 8192) is truncated, not passed through unchanged.
- 100 emoji (200 UTF-16 units) stays unchanged (under 4096).
- Plain-branch parity covered.
- Truncated emoji payload still ends with the escaped ellipsis and every remaining emoji is intact (no orphan surrogate).

---
_Generated by [Claude Code](https://claude.ai/code/session_01FyMxQn2BEAzmwKS1GZczKj)_

---
_Generated by [Claude Code](https://claude.ai/code)_

File: src/chat_sdk/adapters/telegram/format_converter.py [hunk 6]
```diff
@@ -1,45 +1,191 @@
...
+            if not isinstance(item, dict):
+                continue
+            item_children = item.get("children", []) or []
+            content = _render_children(item_children, join="\n")
+            if ordered:
+                marker = escape_markdown_v2(f"{i + 1}.")
+                rendered_items.append(f"{marker} {content}")
+            else:
+                rendered_items.append(f"\\- {content}")
+        return "\n".join(rendered_items)
```
[the remark this answers is not in the document]
Comment (patrick-chinchill) [reply src/chat_sdk/adapters/telegram/format_converter.py:120]: Addressed in commit `00f75f9` ("fix(telegram): preserve ordered-list start value in MarkdownV2 renderer").

The `"list"` branch in `_render_markdown_v2` (see `src/chat_sdk/adapters/telegram/format_converter.py:97-122` on HEAD `f382ea4`) now reads `start = int(node.get("start") or 1)` (with a `try/except` for non-numeric `start`) and uses `enumerate(children, start=start)`. The marker is still escaped via `escape_markdown_v2(f"{i}.")` so the period stays backslash-escaped per MarkdownV2 spec.

Three regression tests in `TestTelegramFromAst`:

- `test_ordered_list_preserves_start_value` -- markdown source `5. five\n6. six\n7. seven` renders as `5\. five` / `6\. six` / `7\. seven` (and no `1\.` / `2\.` / `3\.`).
- `test_ordered_list_with_explicit_start_attribute` -- direct AST input with `start=5` and 3 items renders 5./6./7.
- `test_ordered_list_default_start_is_one` -- omitted `start` still defaults to 1./2.

---
_Generated by [Claude Code](https://claude.ai/code/session_01FyMxQn2BEAzmwKS1GZczKj)_

---
_Generated by [Claude Code](https://claude.ai/code)_

File: tests/test_discord_extended.py [hunk 7]
```diff
@@ -1541,6 +1541,90 @@ async def test_edits_with_card(self):
...
+
+        await adapter.post_message("discord:guild1:channel456", card_message)
+
+        payload = adapter._discord_fetch.call_args[0][2]
+        # Either 'content' is absent entirely, or it's empty/falsy. Both
+        # are acceptable — Discord won't render duplicate text in either
+        # case. The key invariant is that the card body text doesn't show
+        # up as a separate plain message above the embed.
+        assert not payload.get("content")
+        assert "embeds" in payload and len(payload["embeds"]) > 0
```
[the remark this answers is not in the document]
Comment (patrick-chinchill) [reply tests/test_discord_extended.py]: Addressed in commit `f382ea4` ("test(discord): tighten omit-content assertions for card posts").

Both occurrences of `assert not payload.get("content")` (originally at `tests/test_discord_extended.py:1582` and `:1625`) replaced with:

```python
assert "content" not in payload or payload["content"] is None
```

This pins the omit-content contract: `post_message` with a card must omit the `content` key entirely (or set it to `None`) so Discord doesn't render the card body as a separate plain message above the embed. The previous truthiness check would have accepted a regression to `content=""` -- the new assertion fails on that case while still allowing the legitimate `None`/omitted forms.

No production code change needed; the adapter already omits the key correctly (verified by the full-suite run -- 3706 passed). The tests were just lax.

---
_Generated by [Claude Code](https://claude.ai/code/session_01FyMxQn2BEAzmwKS1GZczKj)_

---
_Generated by [Claude Code](https://claude.ai/code)_
